### ISSUE TITLE
Search via Java SDK orders of magnitude slower than REST API

### ISSUE THREAD

**Issue by Splaktar** [splunk/splunk-sdk-java#71]
We have been using the splunk-sdk-java for over a year now. So far we've been very happy with it, but we are starting to run into some performance issues.

This started when we noticed that the minimum time that we could receive a response from our back-end server was around 3 seconds even for the smallest possible search. When kicking off multiple searches serially this becomes a significant delay.

We performed some profiling and discovered that all calls through the splunk-sdk-java were adding a delay of 2.1 to 2.8 seconds in addition to the overhead we have for our own Play controllers. We've since done some experiments with having our back-end use the Splunk REST API instead of the splunk-sdk-java. These have shown orders of magnitude of performance gains. We're seeing responses from the REST API in 100-200ms.

At this point, we're likely going to convert significant portions of our implementation to use the REST API instead of the splunk-sdk-java, but we wanted to report our results so that this could be investigated further.

For further information, our Splunk server is running 5.0.6 and we are using the latest splunk-sdk-java. Splunk is running on Linux and our back-end server is running on Windows. All testing was performed over the local area network over multiple days and time periods.

Thank you.

**Comment by itay** [comment]
@Splaktar thank you very much for filing this. 

First of all, I'm sorry that you're running into this. That type of performance is not right and not something you should be seeing.

In our testing, we haven't seen anything like this before, including on old versions. Given that, I'd really like to get to the bottom of this. We'll set up a local environment to see if we can replicate what you saw.

I have a few questions, and hopefully we can get to the bottom of this:

1. What version of the JVM are you using?
2. I assume you're connecting to a full Enterprise build of Splunk, and not just a Universal Forwarder?

Finally, the thing that would help the most is if you happen to have a program that you've ran that exhibits this slow behavior, so we can do a faithful reproduction locally. If you can't share that, even sharing as complete a code snippet as you can that shows the code you use to create the job would be helpful.

If there is information you are not comfortable sharing here, you can reach us at devinfo@splunk.com and we can take it up there. We'd like to get to the bottom of this ASAP, so let us know how we can help!

**Comment by Splaktar** [comment]
1. I am running 1.7.0_60 and a teammate who is reproducing and debugging this is running 1.6.0.
2. Yes, full Enterprise Splunk, not a Universal Forwarder.
3. The API in use with POST search.

The delay seems to be focused around accessing the search job, either via job creation or other calls that need to build the search job object.

**Comment by itay** [comment]
@Splaktar thanks for the information. Is there any chance you can share a snippet that shows where the issue (i.e. the set of calls that you're seeing 2 seconds of extra time)? Similarly, if you can share the snippet of the direct REST call where you are not seeing it.

Thanks!

**Comment by Splaktar** [comment]
We are working on gathering that information for you now. Thank you for the prompt responses!

**Comment by dojadop** [comment]
The `createJob()` method is okay with response times and is very similar to the REST api.  The key area that we are seeing longer response times is when performing the following:

SDK (500ms -> 2200ms)
`Job job = service.getJobs().get(sid)` 

vs

REST (<80ms)
`GET /services/search/jobs/:sid`

This is also observed using other methods such as `getResults()`, `getResultsPreview()`, etc.

**Comment by itay** [comment]
So if you use the `Job` object returned by `createJob`, and do something like `getResults()`, do you still see a delay there?

**Comment by dojadop** [comment]
When I use the object returned by `createJob` there is no delay.  So that is good news!  However, our use-case for the api doesn't allow us to pass this object around.  We rely on passing the sid to look up the Job and then perform the `getResults()` call.

Any thoughts as to why the get method is having a delay?

**Comment by itay** [comment]
Yes, it has to do with how we list those objects internally.

Generally, when you call `getJobs().get(sid)`, it will fetch all the jobs, and then return them locally. So you're paying for network and then parsing that you are not going to use. There are a couple of reasons why it is done this way, but in your case, it's not giving you good results.

**Comment by itay** [comment]
@Splaktar @dojadop We have a proposed change which you can see here:
https://github.com/splunk/splunk-sdk-java/commit/e32cb40c94a615b33c3724929b234117d674eff3

Would this work for you? It should solve your problem of the API timing.

**Comment by Splaktar** [comment]
That looks to be along the lines of what we're looking for. We'll try it out and report back on the performance. Thank you!

**Comment by glennblock** [comment]
Thanks for being patient with us @Splaktar!

**Comment by dojadop** [comment]
I cloned the feature/getJob branch to test and I am seeing much better response times when I get the `Job` object.  However, after digging into the Job class I noticed every time I access a getter method such as `job.getDoneProgress()` it does an additional request to get this information.  So for the following snippet of code there is still a large latency due to calling the getters sequentially:

```java
Job job = service.getJob(sid);
boolean isDone = job.isDone();       
float doneProgress = job.getDoneProgress();
int resultCount = job.getResultCount();
int resultPreviewCount = job.getResultPreviewCount();
int eventCount = job.getEventCount();
int scanCount = job.getScanCount();
```

The above would add up to be somewhere around 300-400ms depending how many getters were accessed.  Since the Job class implements Map I could just use the get method to access this information locally.  

Also, I was a bit concerned with needing to request all the Jobs before I created one.  So when I perform this call, `service.getJobs().create(search, args)`, will it attempt to go request all the active search jobs due to `service.getJobs()`?

Thanks

**Comment by shakeelmohamed** [comment]
Hi @Splaktar @dojadop,

This is actually a bug in `Job.isReady()`, which is called in `Job.checkReady()` (which is called by `Job.getDoneProgress()` and other `Job.get...` methods).
We've pushed a fix to the same `feature/getJob` branch, see the diff here: 9e66a4dbacc7016da1c730a090c139e10bb5f82a

Thanks again for your patience

**Comment by dojadop** [comment]
Did some more testing with the latest push and seems to be working.  When do you anticipate releasing this fix?

**Comment by shakeelmohamed** [comment]
@Splaktar @dojadop 

We are working on releasing this fix. We'll comment back once it's out along with some other changes coming in the next release.

**Comment by Splaktar** [comment]
Great. Thank you very much!

**Comment by shakeelmohamed** [comment]
You're welcome! We're expecting to get a release out next week as we don't like to release on fridays.

**Comment by shakeelmohamed** [comment]
@Splaktar @dojadop 
We've just released v1.3.2 with these fixes!
Thanks again for reporting this issue and your continued patience.
